0010-endpoint-metadata-extension/README.md

```diff
@@ -0,0 +1,658 @@
+
+  `<endpoint_token>/update/keys`
+
+  where `<endpoint_token>` identifies the endpoint.
+
+The request payload MUST be a JSON-encoded object with the same [JSON schema](http://json-schema.org/) as in [full metadata update request](#full-metadata-update-request).
+In the `metadata` JSON property specify only key-value pairs that need to be updated.
+
+If there is no `id` JSON property in the request payload then server MUST process the request but MUST not publish the status response back to client.
+The server MUST publish the [status response](#partial-metadata-update-response) with the `id` property in the payload and processing status fields.
+
+Examples:
+- update metadata with message identifier
+
+```json
+{
+  "id":42,
+  "metadata":{
+    "deviceModel":"example model",
+    "name":"Sensor 1"
+  }
+}
+```
+- update metadata without message identifier
+
+```json
+{
+  "metadata":{
+    "deviceModel":"example model",
+    "name":"Sensor 1"
+  }
+}
+```
+
+#### Partial metadata update response
+
+The server MUST respond to the metadata get request by publishing the response message according to the [request/response design pattern defined in 1/KP](/0001-kaa-protocol/README.md#requestresponse-pattern).
+The extension-specific resource path part format is:
+
+  `<endpoint_token>/update/keys/status`
+
+The response payload MUST be a JSON-encoded object with the same [JSON schema](http://json-schema.org/) as in [full metadata update request](#full-metadata-update-response).
+
+Examples:
+
+- Endpoint metadata key-value pairs successfully updated
+
+```json
+{
+  "id":42,
+  "statusCode":200,
+  "reasonPhrase":"OK"
+}
+```
+
+### Get metadata
+
+#### Get metadata request
+
+Client MUST send requests with the following extension-specific [resource path](/0001-kaa-protocol/README.md#resource-path-format) part in order to receive the metadata key-value pairs:
+
+  `<endpoint_token>/get`
+
+  where `<endpoint_token>` identifies the endpoint.
+
+The request payload MUST be a JSON-encoded object with the following [JSON schema](http://json-schema.org/) ([file](./schemas/get-metadata-request.schema.json)):
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "10/EPMDX get metadata request",
+  "type": "object",
+  "properties": {
+    "id": {
+      "type": "integer",
+      "description": "The message identifier used to match server response to the request"
+    },
+    "keys": {
+      "type": "array",
+      "items": {
+        "type": "string",
+        "minLength": 1,
+        "description": "The metadata key"
+      },
+      "uniqueItems": true,
+      "description": "The metadata keys to be returned. Optional, if not specified all metadata keys will be returned"
+    }
+  },
+  "additionalProperties": false
+}
+```
+Examples:
+- get metadata without message identifier
```

Comment:
Currently the message identifier in the get request from client is optional and client receives response from server in any case. Should we make it required in the client request?

**Note**: for update/delete operations the message id is also optional and client will not receive the response from server if `"id"` was not present in the request